src/message/kafka/index.ts

```diff
@@ -0,0 +1,249 @@
+  // to the CloudEvents datacontenttype attribute.
+  const headers: Headers = { 
+    ...{ [CONSTANTS.HEADER_CONTENT_TYPE]: event.datacontenttype }, 
+    ...headersFor(event)
+  };
+  return {
+    headers,
+    key: createKey(event),
+    value: event.data,
+    body: event.data,
+    timestamp: timestamp(event.time)
+  };
+}
+
+/**
+ * Serialize a CloudEvent for Kafka in structured mode
+ * @implements {Serializer}
+ * @see https://github.com/cloudevents/spec/blob/v1.0.1/kafka-protocol-binding.md#33-structured-content-mode
+ *
+ * @param {CloudEvent<T>} event the CloudEvent to be serialized
+ * @returns {KafkaMessage<T>} a KafkaMessage instance
+ */
+ function toStructuredKafkaMessage<T>(event: CloudEvent<T>): KafkaMessage<T> {
+  if (event.data_base64) {
+    // The event's data is binary - delete it
+    event = event.cloneWith({ data: undefined });
+  }
+  const value = event.toString();
+  return {
+    key: createKey(event),
+    value,
+    headers: {
+      [CONSTANTS.HEADER_CONTENT_TYPE]: CONSTANTS.DEFAULT_CE_CONTENT_TYPE,
+    },
+    body: value,
+    timestamp: timestamp(event.time)
+  };
+}
+
+/**
+ * Converts a Message to a CloudEvent
+ * @implements {Deserializer}
+ *
+ * @param {Message} message the incoming message
+ * @return {CloudEvent} A new {CloudEvent} instance
+ */
+function deserializeKafkaMessage<T>(message: Message): CloudEvent<T> | CloudEvent<T>[] {
+  if (!isKafkaEvent(message)) {
+    throw new ValidationError("No CloudEvent detected");
+  }
+  if (!(message as KafkaMessage<T>).value) {
+    throw new ValidationError("Value is null or undefined");
+  }
+  if (!message.headers) {
+    throw new ValidationError("Headers are null or undefined");
+  }
+  const cleanHeaders: Headers = sanitize(message.headers);
+  const mode: Mode = getMode(cleanHeaders);
+  switch (mode) {
+    case Mode.BINARY:
+      return parseBinary(message as KafkaMessage<T>);
+    case Mode.STRUCTURED:
+      return parseStructured(message as KafkaMessage<T>);
+    case Mode.BATCH:
+      return parseBatched(message as KafkaMessage<T>);
+    default:
+      throw new ValidationError("Unknown Message mode");
+  }
+}
+
+/**
+ * Determine if a Message is a CloudEvent via Kafka headers
+ * @implements {Detector}
+ *
+ * @param {Message} message an incoming Message object
+ * @returns {boolean} true if this Message is a CloudEvent
+ */
+function isKafkaEvent<T>(message: Message<T>): boolean {
+   const headers = sanitize(message.headers);
+   return !!headers[KAFKA_CE_HEADERS.ID] || // A binary mode event
+    headers[CONSTANTS.HEADER_CONTENT_TYPE]?.startsWith(CONSTANTS.MIME_CE) as boolean || // A structured mode event
+    headers[CONSTANTS.HEADER_CONTENT_TYPE]?.startsWith(CONSTANTS.MIME_CE_BATCH) as boolean; // A batch of events
+}
+
+/**
+ * Creates a Kafka key for the event provided
+ * @param {CloudEvent} e an event
+ * @returns {string} the Kafka key
+ */
+function createKey<T>(e: CloudEvent<T>): string {
+  return `${e.source}/${e.id}`;
```

Comment:
I am glad you used my implementation, but I initially incorrectly implemented the specification.
`Key` cannot be constructed that way. Correct implementation require to use [Partition Key Extenstion](https://github.com/cloudevents/spec/blob/v1.0.1/kafka-protocol-binding.md#31-key-mapping).
I added that extension in [this commit](https://github.com/LeoVS09/cloudevents-js-sdk-kafka/commit/aed9bb89b21b59fa023ef94d94f6054d6004c622)

**It is important to not construct** `key` in such a way not only because of specification but also because this key will be used for partitioning by Kafka. It must be **not unique**. Different events, the order of which is important, must share the same `key`, and still have a unique `id` for correctly processing the Exactly-Once Delivery flow. 
**In simple words:** better to keep this field empty than try to construct it.

Reply:
@LeoVS09 thank you very much for the feedback. I will make the necessary changes for this in the new year. Your implementation has been a very helpful inspiration for this addition and I appreciate both the time you've committed to the code, and that which you spent reviewing the pr. Thanks!